### PR TITLE
Add preemption_target_recomputations_total metric for overlap recomputation outcomes

### PR DESCRIPTION
**Pull request by SatishChoudhary642**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Adds the `kueue_preemption_target_recomputations_total` Prometheus counter metric to track outcomes when preemption targets overlap with targets selected earlier in the same scheduling cycle.

The existing `admission_cycle_preemption_skips` metric is incremented when the scheduler cannot resolve an overlap. However, when recomputation succeeds, the workload can either select alternative targets (`new_targets`) or fit after earlier preemptions complete (`deferred_fit`). This metric provides observability into these successful paths and failed recomputations (`skipped`), enabling operators to see how often the scheduler resolves or fails to resolve overlapping preemption targets.

#### Which issue(s) this PR fixes:
Relates to #14188

#### Special notes for your reviewer:
- Behavior for the existing `admission_cycle_preemption_skips` gauge remains completely unchanged.
- The new counter metric (`kueue_preemption_target_recomputations_total`) includes labels: `cluster_queue`, `result` (which can be `new_targets`, `deferred_fit`, or `skipped`), `replica_role`, and any globally configured custom ClusterQueue metric labels.
- Added unit test cases verifying each of the three outcomes independently in `scheduler_recompute_preemption_test.go`.

#### Does this PR introduce a user-facing change?
```release-note
Observability: Added `kueue_preemption_target_recomputations_total` metric to track the result of overlapping preemption target recomputations within a scheduling cycle.
```

## Summary by CodeRabbit

* **New Features**
  * Added metrics for tracking preemption-target recomputation outcomes by queue and replica role.
  * Reports whether recomputation found new targets, was deferred due to fit constraints, or was skipped.

* **Documentation**
  * Documented the new Prometheus metric, including its labels and possible outcomes.

* **Tests**
  * Added coverage validating recomputation outcomes and corresponding metric values across scheduling scenarios.

### CODE REVIEW

**Comment by SatishChoudhary642** [comment]
@tenzen-y   Can I get lgtm here

File: pkg/metrics/metrics.go
Comment (tenzen-y) [inline pkg/metrics/metrics.go:38]:
```suggestion

type PreemptionTargetRecomputationOutcome string
```

Dedicated typed would be better.

File: pkg/metrics/metrics.go [hunk 1]
```diff
@@ -48,6 +48,16 @@ const (
 	PendingStatusActive       = "active"
 	PendingStatusInadmissible = "inadmissible"
 
+	// PreemptionTargetRecomputationOutcomeNewTargets is recorded when the
+	// recomputation selects non-overlapping preemption targets.
+	PreemptionTargetRecomputationOutcomeNewTargets = "new_targets"
```
Comment (tenzen-y) [inline pkg/metrics/metrics.go]:
```suggestion
	PreemptionTargetRecomputationOutcomeNewTargets PreemptionTargetRecomputationOutcome = "new_targets"
```

File: pkg/metrics/metrics.go [hunk 2]
```diff
@@ -48,6 +48,16 @@ const (
...
+	// PreemptionTargetRecomputationOutcomeDeferredFit is recorded when the
+	// workload will fit only after earlier preemptions in the cycle complete.
+	PreemptionTargetRecomputationOutcomeDeferredFit = "deferred_fit"
```
Comment (tenzen-y) [inline pkg/metrics/metrics.go]:
```suggestion
	PreemptionTargetRecomputationOutcomeDeferredFit PreemptionTargetRecomputationOutcome = "deferred_fit"
```

File: pkg/metrics/metrics.go [hunk 3]
```diff
@@ -48,6 +48,16 @@ const (
...
+	// PreemptionTargetRecomputationOutcomeSkipped is recorded when the
+	// recomputation does not resolve the overlap and the workload is skipped.
+	PreemptionTargetRecomputationOutcomeSkipped = "skipped"
```
Comment (tenzen-y) [inline pkg/metrics/metrics.go]:
```suggestion
	PreemptionTargetRecomputationOutcomeSkipped PreemptionTargetRecomputationOutcome = "skipped"
```

File: pkg/metrics/metrics.go [hunk 4]
```diff
@@ -1459,6 +1487,15 @@ func ReportAdmissionCyclePreemptionSkips(cqName kueue.ClusterQueueReference, cou
 	AdmissionCyclePreemptionSkips.WithLabelValues(labels...).Set(float64(count))
 }
 
+// ReportPreemptionTargetRecomputation increments the counter for a preemption
+// target recomputation outcome. The outcome must be one of
+// PreemptionTargetRecomputationOutcomeNewTargets, PreemptionTargetRecomputationOutcomeDeferredFit,
+// or PreemptionTargetRecomputationOutcomeSkipped.
+func ReportPreemptionTargetRecomputation(cqName kueue.ClusterQueueReference, outcome string, customLabelValues []string, tracker *roletracker.RoleTracker) {
```
Comment (tenzen-y) [inline pkg/metrics/metrics.go]:
```suggestion
func ReportPreemptionTargetRecomputation(cqName kueue.ClusterQueueReference, outcome PreemptionTargetRecomputationOutcome, customLabelValues []string, tracker *roletracker.RoleTracker) {
```

File: pkg/metrics/metrics.go [hunk 5]
```diff
@@ -1459,6 +1487,15 @@ func ReportAdmissionCyclePreemptionSkips(cqName kueue.ClusterQueueReference, cou
...
+	labels := append([]string{string(cqName), outcome, roletracker.GetRole(tracker)}, customLabelValues...)
```
Comment (tenzen-y) [inline pkg/metrics/metrics.go]:
```suggestion
	labels := append([]string{string(cqName), string(outcome), roletracker.GetRole(tracker)}, customLabelValues...)
```

File: pkg/scheduler/scheduler.go [hunk 6]
```diff
@@ -750,7 +753,20 @@ func (s *Scheduler) updateAssignmentIfNeeded(
...
+		switch {
+		case e.assignment.RepresentativeMode() == flavorassigner.DeferredFit:
+			overlapRecomputeOutcome = metrics.PreemptionTargetRecomputationOutcomeDeferredFit
+		case fitsCheck == schdcache.FitsCheckOk && !preemptedWorkloads.HasAny(newTargets):
+			overlapRecomputeOutcome = metrics.PreemptionTargetRecomputationOutcomeNewTargets
+		default:
+			overlapRecomputeOutcome = metrics.PreemptionTargetRecomputationOutcomeSkipped
+		}
+	}
+
```
Comment (tenzen-y) [inline pkg/scheduler/scheduler.go:766]:
```suggestion

	metrics.ReportPreemptionTargetRecomputation(
		e.ClusterQueue,
		outcome,
		s.customLabels.CQGet(e.ClusterQueue),
		s.roleTracker,
)
```

Can we record metric here instead of returning results only for metric recording?

File: pkg/scheduler/scheduler.go [hunk 7]
```diff
@@ -750,7 +753,20 @@ func (s *Scheduler) updateAssignmentIfNeeded(
...
+	return usage, schdcache.FitsCheckOk == fitsCheck, overlapRecomputeOutcome
```
Comment (tenzen-y) [inline pkg/scheduler/scheduler.go]:
```suggestion
	return usage, schdcache.FitsCheckOk == fitsCheck
```

File: pkg/scheduler/scheduler.go [hunk 8]
```diff
@@ -709,7 +712,7 @@ func (s *Scheduler) updateAssignmentIfNeeded(
 	e *entry,
 	snapshot *schdcache.Snapshot,
 	cq *schdcache.ClusterQueueSnapshot,
-	preemptedWorkloads preemption.PreemptedWorkloads) (workload.Usage, bool) {
+	preemptedWorkloads preemption.PreemptedWorkloads) (workload.Usage, bool, string) {
```
Comment (tenzen-y) [inline pkg/scheduler/scheduler.go]:
```suggestion
	preemptedWorkloads preemption.PreemptedWorkloads) (workload.Usage, bool) {
```

File: pkg/scheduler/scheduler_test.go [hunk 9]
```diff
@@ -340,6 +345,20 @@ func runScheduleTestCases(t *testing.T, cfg scheduleTestConfig, cases map[string
...
+
+					for cqName, outcomeWants := range tc.wantPreemptionTargetRecomputations {
+						for outcome, want := range outcomeWants {
+							lvs := []string{cqName, outcome, roletracker.RoleStandalone}
+							val, err := testutil.GetCounterMetricValue(metrics.PreemptionTargetRecomputationsTotal.WithLabelValues(lvs...))
+							if err != nil {
+								t.Fatalf("Couldn't get value for metric preemption_target_recomputations_total for cq=%q outcome=%q: %v", cqName, outcome, err)
+							}
+							got := int(val)
+							if want != got {
```
Comment (tenzen-y) [inline pkg/scheduler/scheduler_test.go]:
```suggestion
							if want != int(val) {
```

nit

File: pkg/scheduler/scheduler.go [hunk 10]
```diff
@@ -419,7 +419,10 @@ func (s *Scheduler) processEntry(
 	// We may also recompute in case of overlapping preemption targets with another workload.
 	// Recompute when needed so CQs considered later in the cycle don't repeatedly
 	// lose to earlier CQs and starve for prolonged periods.
-	usage, fits := s.updateAssignmentIfNeeded(ctx, log, e, snapshot, cq, preemptedWorkloads)
+	usage, fits, overlapRecomputeOutcome := s.updateAssignmentIfNeeded(ctx, log, e, snapshot, cq, preemptedWorkloads)
+	if overlapRecomputeOutcome != "" {
+		metrics.ReportPreemptionTargetRecomputation(e.ClusterQueue, overlapRecomputeOutcome, s.customLabels.CQGet(e.ClusterQueue), s.roleTracker)
+	}
```
Comment (tenzen-y) [inline pkg/scheduler/scheduler.go]:
```suggestion
	usage, fits := s.updateAssignmentIfNeeded(ctx, log, e, snapshot, cq, preemptedWorkloads)
```

**Review by tenzen-y — COMMENTED** [review]
Thank you for working on this, basically LGTM.
Also, can you replace "outcome" word everwhere with "result"?

Because, we used to "result" metric label instead of "outcome" word. We'd like to provide consistency metrics labels.

File: pkg/scheduler/scheduler.go [hunk 11]
```diff
@@ -750,7 +753,20 @@ func (s *Scheduler) updateAssignmentIfNeeded(
...
 	e.NominationMapping = nil
 
-	return usage, schdcache.FitsCheckOk == fitsCheck
+	// Determine the overlap recomputation outcome for metrics reporting.
+	overlapRecomputeOutcome := ""
+	if needsOverlapRecompute {
+		switch {
+		case e.assignment.RepresentativeMode() == flavorassigner.DeferredFit:
+			overlapRecomputeOutcome = metrics.PreemptionTargetRecomputationOutcomeDeferredFit
+		case fitsCheck == schdcache.FitsCheckOk && !preemptedWorkloads.HasAny(newTargets):
```
Comment (tenzen-y) [inline pkg/scheduler/scheduler.go]:
```suggestion
		case len(newTargets) > 0 && fitsCheck == schdcache.FitsCheckOk && !preemptedWorkloads.HasAny(newTargets):
```

If the delayed TAS second-pass wl-A and failed node replacement wl-B nominate the same victim-V as a preemption target, the current condition accidentally selects "new_targets" as an outcome.

File: pkg/metrics/metrics.go
Comment (tenzen-y) [inline pkg/metrics/metrics.go:1246]: Also, I guess that we should clear the new metric here and add `TestReportAndCleanupPreemptionTargetRecomputations` UT in pkg/metrics/metrcis_test.go

**Comment by SatishChoudhary642** [comment]
Thanks so much for the thorough review, @tenzen-y 🫡.
All the suggestions are addressed and pushed.

- Renamed `outcome` → `result` throughout (label key, constants, doc comments, generated metrics table, tests)
- Introduced a dedicated `PreemptionTargetRecomputationResult` type instead of raw strings
- Fixed the `new_targets` edge case — added `len(newTargets) > 0` so an empty target set no longer gets misreported
- Moved the metric recording call inside `updateAssignmentIfNeeded` and reverted its return signature back to `(workload.Usage, bool)`
- Added `PreemptionTargetRecomputationsTotal` cleanup to `ClearClusterQueueMetrics`,

**Comment by tenzen-y** [comment]
/release-note-edit

```release-note
Observability: Added `kueue_preemption_target_recomputations_total` metric to track the result of overlapping preemption target recomputations within a scheduling cycle.
```

**Review by tenzen-y — COMMENTED** [review]
Thank you!
/lgtm
/approve
